### ISSUE TITLE
If resource created with create_before_destroy lifecycle option and after resource creation some of the resources creation fail, terraform still destroys deposed object

### ISSUE THREAD

**Issue by maratsal** [hashicorp/terraform#31635]
If resource created with create_before_destroy lifecycle option and after resource creation some of other resources creation fail, terraform still destroys deposed object.

### Terraform Version

```
Terraform v1.0.8
```

### Terraform Configuration Files

```terraform
provider "google" {
  project = var.project
}

provider "google-beta" {
  project = var.project
}

data "google_container_engine_versions" "default" {
  provider       = google-beta
  for_each       = var.clusters
  location       = each.value.zone
  version_prefix = each.value.version_prefix
}

resource "google_container_cluster" "gke_cluster" {
  for_each = var.clusters

  name     = "${var.env}-ibp-${each.value.name}-cluster"
  project  = var.project
  location = each.value.zone

  min_master_version = data.google_container_engine_versions.default[each.key].latest_node_version
  release_channel {
    channel = "UNSPECIFIED"
  }

  remove_default_node_pool = true
  initial_node_count       = var.initial_node_count

  network    = var.env
  subnetwork = "${var.env}-subnet-${each.value.name}"

  ip_allocation_policy {
    cluster_secondary_range_name  = "${var.env}-subnet-${each.value.name}-sec-01"
    services_secondary_range_name = "${var.env}-subnet-${each.value.name}-sec-02"
  }
}

resource "google_container_node_pool" "node_pool" {
  for_each = var.clusters

  name       = "${var.env}-ibp-${each.value.name}-${random_string.node_pool_name_suffix[each.key].id}"
  project    = var.project
  location   = each.value.zone
  cluster    = google_container_cluster.gke_cluster[each.key].name
  node_count = var.initial_node_count
  version    = data.google_container_engine_versions.default[each.key].latest_node_version

  lifecycle {
    ignore_changes        = [node_count]
    create_before_destroy = true
  }

  node_config {
    preemptible  = var.preemptible_nodes
    machine_type = var.machine_type

    metadata = {
      disable-legacy-endpoints = "true"
    }

    oauth_scopes = [
      "https://www.googleapis.com/auth/logging.write",
      "https://www.googleapis.com/auth/monitoring",
      "https://www.googleapis.com/auth/devstorage.read_only",
    ]
  }

  depends_on = [
    random_string.node_pool_name_suffix,
  ]
}

resource "random_string" "node_pool_name_suffix" {
  for_each = var.clusters
  length   = 6
  special  = false
  upper    = false
  keepers = {
    machine_type = var.machine_type
    preemptible  = var.preemptible_nodes
  }
}

resource "null_resource" "node_pool_provisioner" {
  for_each = var.clusters

  triggers = {
    cluster        = "${var.env}-ibp-${each.value.name}-cluster"
    project        = var.project
    location       = each.value.zone
    node_pool      = "${var.env}-ibp-${each.value.name}-${random_string.node_pool_name_suffix[each.key].id}"
    drain_interval = var.drain_interval
  }

  provisioner "local-exec" {
    #working_dir = "${path.module}/scripts/"
    command = "${path.root}/scripts/drain-nodes.sh ${self.triggers.project} ${self.triggers.location} ${self.triggers.cluster} ${self.triggers.node_pool} ${self.triggers.drain_interval}"
  }

  depends_on = [
    google_container_node_pool.node_pool,
    random_string.node_pool_name_suffix,
  ]
}
```

### Debug Output

```
google_container_node_pool.node_pool["qwerty2"]: Creation complete after 1m11s [id=projects/my-gcp-project/locations/northamerica-northeast2-a/clusters/dev-app-qwerty2-cluster/nodePools/dev-app-qwerty2-0g90ti]
google_container_node_pool.node_pool["qwerty2"] (deposed object d37e2967): Destroying... [id=projects/my-gcp-project/locations/northamerica-northeast2-a/clusters/dev-app-qwerty2-cluster/nodePools/dev-app-qwerty220220630165150246000000001]
google_container_node_pool.node_pool["qwerty1"]: Still creating... [1m20s elapsed]
google_container_node_pool.node_pool["qwerty2"]: Still destroying... [id=projects/my-gcp-project/locations/northame...ev-ibp-qwerty220220630165150246000000001, 10s elapsed]
google_container_node_pool.node_pool["qwerty1"]: Creation complete after 1m22s [id=projects/my-gcp-project/locations/northamerica-northeast1-a/clusters/dev-app-qwerty1-cluster/nodePools/dev-app-qwerty1-lm54zt]
null_resource.node_pool_provisioner["qwerty2"]: Creating...
null_resource.node_pool_provisioner["qwerty1"]: Creating...
null_resource.node_pool_provisioner["qwerty2"]: Provisioning with 'local-exec'...
null_resource.node_pool_provisioner["qwerty2"] (local-exec): Executing: ["/bin/sh" "-c" "scripts/drain-nodes.sh my-gcp-project northamerica-northeast2-a dev-app-qwerty2-cluster dev-app-qwerty2-0g90ti 30"]
null_resource.node_pool_provisioner["qwerty2"] (local-exec): /bin/sh: scripts/drain-nodes.sh: not found
null_resource.node_pool_provisioner["qwerty1"]: Provisioning with 'local-exec'...
null_resource.node_pool_provisioner["qwerty1"] (local-exec): Executing: ["/bin/sh" "-c" "scripts/drain-nodes.sh my-gcp-project northamerica-northeast1-a dev-app-qwerty1-cluster dev-app-qwerty1-lm54zt 30"]
null_resource.node_pool_provisioner["qwerty1"] (local-exec): /bin/sh: scripts/drain-nodes.sh: not found
google_container_node_pool.node_pool["qwerty1"] (deposed object dfe140b0): Destroying... [id=projects/my-gcp-project/locations/northamerica-northeast1-a/clusters/dev-app-qwerty1-cluster/nodePools/dev-app-qwerty120220630165150254900000002]
google_container_node_pool.node_pool["qwerty2"]: Still destroying... [id=projects/my-gcp-project/locations/northame...ev-ibp-qwerty220220630165150246000000001, 20s elapsed]
```

### Expected Behavior

I would expect terraform run to stop right on failed resource without deposed resource deletion

### Actual Behavior

Terraform destroyed deposed resource after another resource creation has failed.

### Steps to Reproduce

```
terraform init
terraform apply
```

### Additional Context

N/A

### References

N/A

**Comment by jbardin** [comment]
Hi @maratsal,

Thanks for filing the issue. I cannot explain the ordering you have show here. The expected order of operations when `node_pool_provisioner` and `gke_cluster` are replaced is exactly the same as the example [here](https://github.com/hashicorp/terraform/blob/main/docs/destroying.md#create-before-destroy). Just to double check, creating a minimal example with the same dependencies also works as you would expect, with the deposed nodes not being executed until after the provisioners have completed.

I don't recall any related bugs since then, but you are using an older version of terraform which could change things. Can you reproduce this with a current release?

Is it possible you made some configuration changes before trying to replace the resource which may have affected the dependencies?

Thanks!

**Comment by jbardin** [comment]
Hi @maratsal,

I have managed to replicate the behavior by corrupting the dependency information stored in the terraform state file. Is it possible that the state for these resources was created with an earlier version of terraform, and no apply was done during the upgrade process to update the state data?

**Comment by corey-hammerton** [comment]
Hi @maratsal. 

I can reproduce this when using [google_compute_instance_template](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance_template), with `lifecycle.create_before_destroy` configured, and [google_storage_bucket_object](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_object). The former refers to the latter for `metadata.startup-script-url`. 

I can see the results of a `terraform plan` show only these resource types marked as `+/- create replacement and then destroy`

**Comment by jbardin** [comment]
Hi @corey-hammerton, what you are describing is a different situation from this issue, which is about deposed instances not being destroy in an order relative to certain dependencies. In your case you have resources which must be replace with the `create_before_destroy` ordering because there are downstream dependencies which require that order.

**Comment by maratsal** [comment]
I was trying to reproduce the issue and failed to do so and I believe I know what has happened.

Right before this apply (where I had failure) I introduced modifications to the module:
- added dependency to the module `google_container_node_pool` 
- added `null_resource` with dependencies
- added `random_string`

I think because of that on failure of null_resource module terraform didn't stop destruction of the deposed `google_container_node_pool` resource.

I will test this out to confirm and will let you know.

Sorry for caused confusion with this  bug report.

**Comment by maratsal** [comment]
I could reproduce the issue.

Module before changes:
```
provider "google" {
  project = var.project
}

provider "google-beta" {
  project = var.project
}

data "google_container_engine_versions" "default" {
  provider       = google-beta
  for_each       = var.clusters
  location       = each.value.zone
  version_prefix = each.value.version_prefix
}

resource "google_container_cluster" "gke_cluster" {
  for_each = var.clusters

  name     = "${var.env}-ibp-${each.value.name}-cluster"
  project  = var.project
  location = each.value.zone

  min_master_version = data.google_container_engine_versions.default[each.key].latest_node_version
  release_channel {
    channel = "UNSPECIFIED"
  }

  remove_default_node_pool = true
  initial_node_count       = var.initial_node_count

  network    = var.env
  subnetwork = "${var.env}-subnet-${each.value.name}"

  ip_allocation_policy {
    cluster_secondary_range_name  = "${var.env}-subnet-${each.value.name}-sec-01"
    services_secondary_range_name = "${var.env}-subnet-${each.value.name}-sec-02"
  }
}

resource "google_container_node_pool" "node_pool" {
  for_each = var.clusters

  name_prefix = "${var.env}-ibp-${each.value.name}-"
  project    = var.project
  location   = each.value.zone
  cluster    = google_container_cluster.gke_cluster[each.key].name
  node_count = var.initial_node_count
  version    = data.google_container_engine_versions.default[each.key].latest_node_version

  lifecycle {
    ignore_changes        = [node_count]
    create_before_destroy = true
  }

  node_config {
    preemptible  = var.preemptible_nodes
    machine_type = var.machine_type

    metadata = {
      disable-legacy-endpoints = "true"
    }

    oauth_scopes = [
      "https://www.googleapis.com/auth/logging.write",
      "https://www.googleapis.com/auth/monitoring",
      "https://www.googleapis.com/auth/devstorage.read_only",
    ]
  }
}
```

Module after changes:
```
provider "google" {
  project = var.project
}

provider "google-beta" {
  project = var.project
}

data "google_container_engine_versions" "default" {
  provider       = google-beta
  for_each       = var.clusters
  location       = each.value.zone
  version_prefix = each.value.version_prefix
}

resource "google_container_cluster" "gke_cluster" {
  for_each = var.clusters

  name     = "${var.env}-ibp-${each.value.name}-cluster"
  project  = var.project
  location = each.value.zone

  min_master_version = data.google_container_engine_versions.default[each.key].latest_node_version
  release_channel {
    channel = "UNSPECIFIED"
  }

  remove_default_node_pool = true
  initial_node_count       = var.initial_node_count

  network    = var.env
  subnetwork = "${var.env}-subnet-${each.value.name}"

  ip_allocation_policy {
    cluster_secondary_range_name  = "${var.env}-subnet-${each.value.name}-sec-01"
    services_secondary_range_name = "${var.env}-subnet-${each.value.name}-sec-02"
  }
}

resource "google_container_node_pool" "node_pool" {
  for_each = var.clusters

  name       = "${var.env}-ibp-${each.value.name}-${random_string.node_pool_name_suffix[each.key].id}"
  project    = var.project
  location   = each.value.zone
  cluster    = google_container_cluster.gke_cluster[each.key].name
  node_count = var.initial_node_count
  version    = data.google_container_engine_versions.default[each.key].latest_node_version

  lifecycle {
    ignore_changes        = [node_count]
    create_before_destroy = true
  }

  node_config {
    preemptible  = var.preemptible_nodes
    machine_type = var.machine_type

    metadata = {
      disable-legacy-endpoints = "true"
    }

    oauth_scopes = [
      "https://www.googleapis.com/auth/logging.write",
      "https://www.googleapis.com/auth/monitoring",
      "https://www.googleapis.com/auth/devstorage.read_only",
    ]
  }

  depends_on = [
    random_string.node_pool_name_suffix,
  ]
}

resource "random_string" "node_pool_name_suffix" {
  for_each = var.clusters
  length   = 6
  special  = false
  upper    = false
  keepers = {
    machine_type = var.machine_type
    preemptible  = var.preemptible_nodes
  }
}

resource "null_resource" "node_pool_provisioner" {
  for_each = var.clusters

  triggers = {
    cluster        = "${var.env}-ibp-${each.value.name}-cluster"
    project        = var.project
    location       = each.value.zone
    node_pool      = "${var.env}-ibp-${each.value.name}-${random_string.node_pool_name_suffix[each.key].id}"
    drain_interval = var.drain_interval
  }

  provisioner "local-exec" {
    #working_dir = "${path.module}/scripts/"
    command = "${path.root}/scripts/drain-nodes.sh ${self.triggers.project} ${self.triggers.location} ${self.triggers.cluster} ${self.triggers.node_pool} ${self.triggers.drain_interval}"
  }

  depends_on = [
    google_container_node_pool.node_pool,
    random_string.node_pool_name_suffix,
  ]
}
```

Terraform apply with failure after module changes:
```
Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

  # google_container_cluster.gke_cluster["reg1"] has been changed
  ~ resource "google_container_cluster" "gke_cluster" {
        id                          = "projects/my-nice-project/locations/northamerica-northeast1-a/clusters/my-nice-prefix-cluster"
        name                        = "my-nice-prefix-cluster"
      + resource_labels             = {}
        # (26 unchanged attributes hidden)

      + node_config {
          + disk_size_gb      = 100
          + disk_type         = "pd-standard"
          + guest_accelerator = []
          + image_type        = "COS_CONTAINERD"
          + labels            = {}
          + local_ssd_count   = 0
          + machine_type      = "e2-medium"
          + metadata          = {
              + "disable-legacy-endpoints" = "true"
            }
          + oauth_scopes      = [
              + "https://www.googleapis.com/auth/devstorage.read_only",
              + "https://www.googleapis.com/auth/logging.write",
              + "https://www.googleapis.com/auth/monitoring",
            ]
          + preemptible       = false
          + service_account   = "default"
          + spot              = false
          + tags              = []
          + taint             = []

          + shielded_instance_config {
              + enable_integrity_monitoring = true
              + enable_secure_boot          = false
            }
        }

      + node_pool {
          + initial_node_count          = 3
          + instance_group_urls         = [
              + "https://www.googleapis.com/compute/v1/projects/my-nice-project/zones/northamerica-northeast1-a/instanceGroupManagers/gke-my-nice-prefix-cl-my-nice-prefix-20-2bc07a60-grp",
            ]
          + managed_instance_group_urls = [
              + "https://www.googleapis.com/compute/v1/projects/my-nice-project/zones/northamerica-northeast1-a/instanceGroups/gke-my-nice-prefix-cl-my-nice-prefix-20-2bc07a60-grp",
            ]
          + max_pods_per_node           = 110
          + name                        = "my-nice-prefix-20220817151912534300000001"
          + node_count                  = 3
          + node_locations              = [
              + "northamerica-northeast1-a",
            ]
          + version                     = "1.22.8-gke.2200"

          + management {
              + auto_repair  = true
              + auto_upgrade = true
            }

          + node_config {
              + disk_size_gb      = 100
              + disk_type         = "pd-standard"
              + guest_accelerator = []
              + image_type        = "COS_CONTAINERD"
              + labels            = {}
              + local_ssd_count   = 0
              + machine_type      = "e2-medium"
              + metadata          = {
                  + "disable-legacy-endpoints" = "true"
                }
              + oauth_scopes      = [
                  + "https://www.googleapis.com/auth/devstorage.read_only",
                  + "https://www.googleapis.com/auth/logging.write",
                  + "https://www.googleapis.com/auth/monitoring",
                ]
              + preemptible       = false
              + service_account   = "default"
              + spot              = false
              + tags              = []
              + taint             = []

              + shielded_instance_config {
                  + enable_integrity_monitoring = true
                  + enable_secure_boot          = false
                }
            }

          + upgrade_settings {
              + max_surge       = 1
              + max_unavailable = 0
            }
        }


        # (12 unchanged blocks hidden)
    }
  # google_container_node_pool.node_pool["reg1"] has been changed
  ~ resource "google_container_node_pool" "node_pool" {
        id                          = "projects/my-nice-project/locations/northamerica-northeast1-a/clusters/my-nice-prefix-cluster/nodePools/my-nice-prefix-20220817151912534300000001"
        name                        = "my-nice-prefix-20220817151912534300000001"
        # (11 unchanged attributes hidden)


      ~ node_config {
          + tags              = []
            # (13 unchanged attributes hidden)

            # (1 unchanged block hidden)
        }

        # (2 unchanged blocks hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the
relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create
+/- create replacement and then destroy

Terraform will perform the following actions:

  # google_container_node_pool.node_pool["reg1"] must be replaced
+/- resource "google_container_node_pool" "node_pool" {
      ~ id                          = "projects/my-nice-project/locations/northamerica-northeast1-a/clusters/my-nice-prefix-cluster/nodePools/my-nice-prefix-20220817151912534300000001" -> (known after apply)
      ~ initial_node_count          = 3 -> (known after apply)
      ~ instance_group_urls         = [
          - "https://www.googleapis.com/compute/v1/projects/my-nice-project/zones/northamerica-northeast1-a/instanceGroupManagers/gke-my-nice-prefix-cl-my-nice-prefix-20-2bc07a60-grp",
        ] -> (known after apply)
      ~ managed_instance_group_urls = [
          - "https://www.googleapis.com/compute/v1/projects/my-nice-project/zones/northamerica-northeast1-a/instanceGroups/gke-my-nice-prefix-cl-my-nice-prefix-20-2bc07a60-grp",
        ] -> (known after apply)
      ~ max_pods_per_node           = 110 -> (known after apply)
      ~ name                        = "my-nice-prefix-20220817151912534300000001" -> (known after apply)
      ~ node_locations              = [
          - "northamerica-northeast1-a",
        ] -> (known after apply)
      + operation                   = (known after apply)
        # (6 unchanged attributes hidden)

      ~ management {
          ~ auto_repair  = true -> (known after apply)
          ~ auto_upgrade = true -> (known after apply)
        }

      ~ node_config {
          ~ disk_size_gb      = 100 -> (known after apply)
          ~ disk_type         = "pd-standard" -> (known after apply)
          ~ guest_accelerator = [] -> (known after apply)
          ~ image_type        = "COS_CONTAINERD" -> (known after apply)
          ~ labels            = {} -> (known after apply)
          ~ local_ssd_count   = 0 -> (known after apply)
          ~ machine_type      = "e2-medium" -> "e2-standard-4" # forces replacement
          + min_cpu_platform  = (known after apply)
          ~ service_account   = "default" -> (known after apply)
          - tags              = [] -> null
          ~ taint             = [] -> (known after apply)
            # (4 unchanged attributes hidden)

          ~ shielded_instance_config {
              ~ enable_integrity_monitoring = true -> (known after apply)
              ~ enable_secure_boot          = false -> (known after apply)
            }

          + workload_metadata_config {
              + mode = (known after apply)
            }
        }

      ~ upgrade_settings {
          ~ max_surge       = 1 -> (known after apply)
          ~ max_unavailable = 0 -> (known after apply)
        }
    }

  # null_resource.node_pool_provisioner["reg1"] will be created
  + resource "null_resource" "node_pool_provisioner" {
      + id       = (known after apply)
      + triggers = (known after apply)
    }

  # random_string.node_pool_name_suffix["reg1"] will be created
  + resource "random_string" "node_pool_name_suffix" {
      + id          = (known after apply)
      + keepers     = {
          + "machine_type" = "e2-standard-4"
          + "preemptible"  = "false"
        }
      + length      = 6
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + numeric     = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

Plan: 3 to add, 0 to change, 1 to destroy.

random_string.node_pool_name_suffix["reg1"]: Creating...
random_string.node_pool_name_suffix["reg1"]: Creation complete after 0s [id=eolg0x]
google_container_node_pool.node_pool["reg1"]: Creating...
google_container_node_pool.node_pool["reg1"]: Still creating... [10s elapsed]
google_container_node_pool.node_pool["reg1"]: Still creating... [20s elapsed]
google_container_node_pool.node_pool["reg1"]: Still creating... [30s elapsed]
google_container_node_pool.node_pool["reg1"]: Still creating... [40s elapsed]
google_container_node_pool.node_pool["reg1"]: Still creating... [50s elapsed]
google_container_node_pool.node_pool["reg1"]: Still creating... [1m0s elapsed]
google_container_node_pool.node_pool["reg1"]: Still creating... [1m10s elapsed]
google_container_node_pool.node_pool["reg1"]: Creation complete after 1m12s [id=projects/my-nice-project/locations/northamerica-northeast1-a/clusters/my-nice-prefix-cluster/nodePools/my-nice-prefix-20220817160943543600000001]
null_resource.node_pool_provisioner["reg1"]: Creating...
google_container_node_pool.node_pool["reg1"] (deposed object 840966d4): Destroying... [id=projects/my-nice-project/locations/northamerica-northeast1-a/clusters/my-nice-prefix-cluster/nodePools/my-nice-prefix-20220817151912534300000001]
null_resource.node_pool_provisioner["reg1"]: Provisioning with 'local-exec'...
null_resource.node_pool_provisioner["reg1"] (local-exec): Executing: ["/bin/sh" "-c" "./scripts/drain-nodes.sh my-nice-project northamerica-northeast1-a my-nice-prefix-cluster my-nice-prefix-eolg0x 30"]
null_resource.node_pool_provisioner["reg1"] (local-exec): /bin/sh: ./scripts/drain-nodes.sh: not found
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 10s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 20s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 30s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 40s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 50s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 1m0s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 1m10s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 1m20s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 1m30s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 1m40s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 1m50s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 2m0s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 2m10s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 2m20s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 2m30s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 2m40s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 2m50s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 3m0s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 3m10s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 3m20s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 3m30s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 3m40s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 3m50s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 4m0s elapsed]
google_container_node_pool.node_pool["reg1"]: Still destroying... [id=projects/my-nice-project/locations/no...refix-reg1-20220817151912534300000001, 4m10s elapsed]
google_container_node_pool.node_pool["reg1"]: Destruction complete after 4m11s
╷
│ Error: local-exec provisioner error
│ 
│   with null_resource.node_pool_provisioner["reg1"],
│   on main.tf line 98, in resource "null_resource" "node_pool_provisioner":
│   98:   provisioner "local-exec" {
│ 
│ Error running command './scripts/drain-nodes.sh my-nice-project
│ northamerica-northeast1-a my-nice-prefix-cluster my-nice-prefix-eolg0x 30':
│ exit status 127. Output: /bin/sh: ./scripts/drain-nodes.sh: not found
│ 
```

**Comment by jbardin** [comment]
Here's a little more condensed reproduction case
```
variable "rev" {
  type = number
}

resource "null_resource" "node_pool_provisioner" {
  count = var.rev
  triggers = {
    node_pool = null_resource.node_pool.id
  }

  provisioner "local-exec" {
    command = "exit 1"
  }
}

resource "null_resource" "node_pool" {
  triggers = {
    name = "${var.rev}"
  }

  lifecycle {
    create_before_destroy = true
  }
}
```


```
% terraform apply -auto-approve -var rev=0
% terraform apply -auto-approve -var rev=1
```

**Comment by jbardin** [comment]
I'm not entirely certain that a solution here is possible which doesn't require some sort of manual intervention by the user, but the current situation where it silently fails in the incorrect order is definitely not the best.

To lay out some of the complexity here:

Destroying is always done using the dependencies stored in the state for that instance. This reduces the likelihood that configuration changes which alter dependencies result in impossible ordering for the destroy operation. The problem here (using the minimal example above) arises because the `null_resource.node_pool` instance does not have any stored dependencies on `null_resource.node_pool_provisioner` because there were no instances a the time the `null_resource.node_pool` was created.

Since trying to insert dependencies from the current config into the prior instance state is bound to cause some unexpected failures, I think the best we can do is try to detect the situation ahead of time. This means that adding an entirely new `null_resource.node_pool_provisioner` to the config at the same time some other resource were replaced would still fail in the same way, but that feels less surprising since it's more obvious that the `null_resource.node_pool_provisioner` did not exists in any way previously. It may then be possible to use the empty `null_resource.node_pool_provisioner` expansion node to detect dependencies rather than the instance nodes, so that it gets recorded in the state. If that's possible, it would work out just fine during destroy, since dependencies are tracked by whole resource addresses, and dependencies are not actually required to be present during destroy at all.

**Comment by jbardin** [comment]
OK, as it turns out the problem is a little trickier still, the "dependency" here required for `create_before_destroy` of course isn't part of the `null_resource.node_pool`, it's the reference from `null_resource.node_pool_provisioner` which is reversed by `create_before_destroy`. When building the apply graph, there are not yet any `null_resource.node_pool_provisioner` instances to connect to the `null_resource.node_pool`, which is why the dependency can't be taken into account.

It may be possible that the expansion node could return all references from its entire resource block so the edges can be connected to that node as well, which would establish the correct edges in the graph.

**Comment by jbardin** [comment]
This was mostly fixed back in v1.4. Mostly meaning that it will pickup the configuration dependency only if there are no stored dependencies in the state (e.g. my minimal example above fixed). If there are stored dependencies, then we still abide by those for the reasons mentioned above. Given that this is already a very narrow edge case, and there is no solution without various tradeoffs leaving the user to manually fix it in many cases no matter what, I think we can close this with the current behavior.
